[GPU Process][FormControls] Don't create a fake NSView if viewless cell drawing is supported

shallawa opened the pull request:
#### d9b43d246262ba0832806f7aa7dcb6e6fbc4806f
<pre>
[GPU Process][FormControls] Don&apos;t create a fake NSView if viewless cell drawing is supported
<a href="https://bugs.webkit.org/show_bug.cgi?id=252723">https://bugs.webkit.org/show_bug.cgi?id=252723</a>
rdar://105778966

Reviewed by Aditya Keerthi.

Currently, we create an NSView unconditionally which may or may not be used.
Creating the fake NSView should be restricted to the case when the viewless
drawing is not supported.

Because viewless cell drawing is still in progress, we are going to enable it if
GPUP is enabled on macOS only for now.

This will require moving the call to ControlFactoryMac::drawingView() from the
super classes `draw()` methods deeper inside the base class ControlMac methods
after we check `supportsViewlessCells()` and find it is false.

The call to `[window setDefaultButtonCell:buttonCell]` was removed from drawing
the default button. We call `[m_defaultButtonCell setKeyEquivalent:@&quot;\r&quot;]` when
creating the NSButtonCell for the default button. And this is sufficient to make
the drawing correct: the dark background and the white text.

The calls to [sliderCell startTrackingAt:] and [sliderCell stopTracking:] were
removed also. The methods seem to be related to interacting with NSSliderCell
and they do not affect the drawing. The interaction with the SliderThumb part is
managed by WebKit which decides where to display this control part.

ControlMac::drawCellOrFocusRing() should be responsible of creating an instance
of LocalCurrentGraphicsContext because the caller which is `ControlMac::drawCell()`
may change the drawing GraphicsContext to an intermediate ImageBuffer for scaling.

For clarity, make instantiating an instance of LocalDefaultSystemAppearance and
an instance of GraphicsContextStateSaver if needed be the responsibility of the
super classes `draw()` methods.

Another clean-up was added. The exception block controls can be moved from the
super classes `draw()` methods to ControlMac::drawCellOrFocusRing().

* Source/WebCore/platform/graphics/mac/controls/ButtonMac.mm:
(WebCore::ButtonMac::draw):
* Source/WebCore/platform/graphics/mac/controls/ColorWellMac.mm:
(WebCore::ColorWellMac::draw):
* Source/WebCore/platform/graphics/mac/controls/ControlMac.h:
* Source/WebCore/platform/graphics/mac/controls/ControlMac.mm:
(WebCore::supportsViewlessCells):
(WebCore::drawViewlessCell):
(WebCore::ControlMac::drawCellInternal):
(WebCore::drawViewlessCellFocusRing):
(WebCore::ControlMac::drawCellFocusRingInternal):
(WebCore::ControlMac::drawCellFocusRing):
(WebCore::ControlMac::drawCellOrFocusRing):
(WebCore::ControlMac::drawCell):
(WebCore::ControlMac::drawListButton):
(WebCore::drawCellFocusRing): Deleted.
* Source/WebCore/platform/graphics/mac/controls/ImageControlsButtonMac.mm:
(WebCore::ImageControlsButtonMac::draw):
* Source/WebCore/platform/graphics/mac/controls/MenuListMac.mm:
(WebCore::MenuListMac::draw):
* Source/WebCore/platform/graphics/mac/controls/MeterMac.mm:
(WebCore::MeterMac::draw):
* Source/WebCore/platform/graphics/mac/controls/SearchFieldCancelButtonMac.mm:
(WebCore::SearchFieldCancelButtonMac::draw):
* Source/WebCore/platform/graphics/mac/controls/SearchFieldMac.mm:
(WebCore::SearchFieldMac::draw):
* Source/WebCore/platform/graphics/mac/controls/SearchFieldResultsMac.mm:
(WebCore::SearchFieldResultsMac::draw):
* Source/WebCore/platform/graphics/mac/controls/SliderThumbMac.mm:
(WebCore::SliderThumbMac::updateCellStates):
(WebCore::SliderThumbMac::draw):
* Source/WebCore/platform/graphics/mac/controls/SliderTrackMac.mm:
(WebCore::SliderTrackMac::draw):
* Source/WebCore/platform/graphics/mac/controls/TextFieldMac.mm:
(WebCore::TextFieldMac::draw):
* Source/WebCore/platform/graphics/mac/controls/ToggleButtonMac.mm:
(WebCore::ToggleButtonMac::draw):

Canonical link: <a href="https://commits.webkit.org/260726@main">https://commits.webkit.org/260726@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a3673add295a5a259cc259bcc96006f8629f5b61

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/109164 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/18243 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/41977 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/691 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/118405 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/113046 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/19702 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/9545 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/101404 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/114922 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/14773 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/98004 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/42955 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/96745 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/29658 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/84677 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/11022 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/31001 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/11751 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/7934 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/17120 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/50604 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7400 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/13369 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/platform/graphics/mac/controls/ControlMac.mm:
```diff
@@ -242,12 +251,50 @@ static void drawCellInView(const FloatRect& rect, NSCell *cell, NSView *view)
+{
+    applyViewlessCellSettings(deviceScaleFactor, style, cell);
+
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wnonnull"
+    [cell drawFocusRingMaskWithFrame:rect inView:nil];
+#pragma clang diagnostic pop
```
Do we have an AppKit radar to allow us to pass `nil` here?

shallawa replied:
Yes. It is rdar://105250010. A FIXME comment was added.

pxlcoder commented on Source/WebCore/platform/graphics/mac/controls/ControlMac.h:
```diff
@@ -63,8 +63,11 @@ class ControlMac : public PlatformControl {
-    void drawCell(GraphicsContext&, const FloatRect&, float deviceScaleFactor, const ControlStyle&, NSCell *, NSView *, bool drawCell = true);
-    void drawCellOrFocusRing(GraphicsContext&, const FloatRect&, float deviceScaleFactor, const ControlStyle&, NSCell *, NSView *, bool drawCell = true);
+    void drawCellInternal(const FloatRect&, float deviceScaleFactor, const ControlStyle&, NSCell *);
+    void drawCellFocusRingInternal(const FloatRect&, float deviceScaleFactor, const ControlStyle&, NSCell *);
```
Can these "Internal" methods, be `private` rather than `protected`?

shallawa replied:
Fixed. And I made `TextFieldMac::draw()` call `drawCell()` instead of calling `drawCellInternal()`.

pxlcoder commented on Source/WebCore/platform/graphics/mac/controls/TextFieldMac.mm:
```diff
@@ -78,11 +78,8 @@
         [m_textFieldCell.get() setEnabled:enabled];
-        [m_textFieldCell.get() drawWithFrame:NSRect(paintRect) inView:view];
-        [m_textFieldCell.get() setControlView:nil];
+        drawCellInternal(paintRect, deviceScaleFactor, style, m_textFieldCell.get());
```
Why is `drawCellInternal` called directly here?

pxlcoder commented on Source/WebCore/platform/graphics/mac/controls/SliderThumbMac.mm:
```diff
@@ -72,12 +64,11 @@
 void SliderThumbMac::draw(GraphicsContext& context, const FloatRoundedRect& borderRect, float deviceScaleFactor, const ControlStyle& style)
 {
     LocalDefaultSystemAppearance localAppearance(style.states.contains(ControlStyle::State::DarkAppearance), style.accentColor);
-    LocalCurrentGraphicsContext localContext(context);
```
Is this safe to remove? We're still transforming the context below (L74).

shallawa replied:
`GraphicsContextStateSaver` will be kept. I made `LocalCurrentGraphicsContext` be instantiated by `ControlMac::drawCellOrFocusRing()` which is called by `ControlMac::drawCell()` because the drawing `GraphicsContext` can be switched to an intermediate `ImageBuffer`.

pxlcoder commented on Source/WebCore/platform/graphics/mac/controls/InnerSpinButtonMac.mm:
```diff
@@ -90,8 +89,7 @@
         context.scale(style.zoomFactor);
     }
 
-    LocalCurrentGraphicsContext localContext(context);
```
Was this just always unnecessary?

shallawa replied:
I think yes. `[NSAppearance _drawInRect:]` is just a wrapper for a CoreUI method which takes a `CGContextRef` and does not rely on `[NSGraphicsContext currentContext]`.

pxlcoder commented on Source/WebCore/platform/graphics/mac/controls/ControlMac.mm:
```diff
@@ -325,7 +367,6 @@ static void drawCellFocusRing(const FloatRect& rect, NSCell *cell, NSView *view)
         return;
 
     // We can't paint an NSComboBoxCell since they are not height-resizable.
-    LocalDefaultSystemAppearance localAppearance(style.states.contains(ControlStyle::State::DarkAppearance), style.accentColor);
```
Note: I think this is safe to remove, since the callers of `drawListButton` set the system appearance.

shallawa replied:
Yes this is right.

shallawa committed:
[GPU Process][FormControls] Don't create a fake NSView if viewless cell drawing is supported
https://bugs.webkit.org/show_bug.cgi?id=252723
rdar://105778966

Reviewed by Aditya Keerthi.

Currently, we create an NSView unconditionally which may or may not be used.
Creating the fake NSView should be restricted to the case when the viewless
drawing is not supported.

Because viewless cell drawing is still in progress, we are going to enable it if
GPUP is enabled on macOS only for now.

This will require moving the call to ControlFactoryMac::drawingView() from the
super classes `draw()` methods deeper inside the base class ControlMac methods
after we check `supportsViewlessCells()` and find it is false.

The call to `[window setDefaultButtonCell:buttonCell]` was removed from drawing
the default button. We call `[m_defaultButtonCell setKeyEquivalent:@"\r"]` when
creating the NSButtonCell for the default button. And this is sufficient to make
the drawing correct: the dark background and the white text.

The calls to [sliderCell startTrackingAt:] and [sliderCell stopTracking:] were
removed also. The methods seem to be related to interacting with NSSliderCell
and they do not affect the drawing. The interaction with the SliderThumb part is
managed by WebKit which decides where to display this control part.

ControlMac::drawCellOrFocusRing() should be responsible of creating an instance
of LocalCurrentGraphicsContext because the caller which is `ControlMac::drawCell()`
may change the drawing GraphicsContext to an intermediate ImageBuffer for scaling.

For clarity, make instantiating an instance of LocalDefaultSystemAppearance and
an instance of GraphicsContextStateSaver if needed be the responsibility of the
super classes `draw()` methods.

Another clean-up was added. The exception block controls can be moved from the
super classes `draw()` methods to ControlMac::drawCellOrFocusRing().

* Source/WebCore/platform/graphics/mac/controls/ButtonMac.mm:
(WebCore::ButtonMac::draw):
* Source/WebCore/platform/graphics/mac/controls/ColorWellMac.mm:
(WebCore::ColorWellMac::draw):
* Source/WebCore/platform/graphics/mac/controls/ControlMac.h:
* Source/WebCore/platform/graphics/mac/controls/ControlMac.mm:
(WebCore::supportsViewlessCells):
(WebCore::drawViewlessCell):
(WebCore::ControlMac::drawCellInternal):
(WebCore::drawViewlessCellFocusRing):
(WebCore::ControlMac::drawCellFocusRingInternal):
(WebCore::ControlMac::drawCellFocusRing):
(WebCore::ControlMac::drawCellOrFocusRing):
(WebCore::ControlMac::drawCell):
(WebCore::ControlMac::drawListButton):
(WebCore::drawCellFocusRing): Deleted.
* Source/WebCore/platform/graphics/mac/controls/ImageControlsButtonMac.mm:
(WebCore::ImageControlsButtonMac::draw):
* Source/WebCore/platform/graphics/mac/controls/MenuListMac.mm:
(WebCore::MenuListMac::draw):
* Source/WebCore/platform/graphics/mac/controls/MeterMac.mm:
(WebCore::MeterMac::draw):
* Source/WebCore/platform/graphics/mac/controls/SearchFieldCancelButtonMac.mm:
(WebCore::SearchFieldCancelButtonMac::draw):
* Source/WebCore/platform/graphics/mac/controls/SearchFieldMac.mm:
(WebCore::SearchFieldMac::draw):
* Source/WebCore/platform/graphics/mac/controls/SearchFieldResultsMac.mm:
(WebCore::SearchFieldResultsMac::draw):
* Source/WebCore/platform/graphics/mac/controls/SliderThumbMac.mm:
(WebCore::SliderThumbMac::updateCellStates):
(WebCore::SliderThumbMac::draw):
* Source/WebCore/platform/graphics/mac/controls/SliderTrackMac.mm:
(WebCore::SliderTrackMac::draw):
* Source/WebCore/platform/graphics/mac/controls/TextFieldMac.mm:
(WebCore::TextFieldMac::draw):
* Source/WebCore/platform/graphics/mac/controls/ToggleButtonMac.mm:
(WebCore::ToggleButtonMac::draw):

Canonical link: https://commits.webkit.org/260726@main
